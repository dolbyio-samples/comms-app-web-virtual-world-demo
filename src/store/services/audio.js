import VoxeetSDK from "@voxeet/voxeet-web-sdk";

export const startRemoteAudio = (participant) => {
  return VoxeetSDK.audio.remote
    .start(participant)
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopRemoteAudio = (participant) => {
  return VoxeetSDK.audio.remote
    .stop(participant)
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const startLocalAudio = () => {
  return VoxeetSDK.audio.local
    .start()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};

export const stopLocalAudio = () => {
  return VoxeetSDK.audio.local
    .stop()
    .then((response) => ({ response }))
    .catch((error) => ({ error }));
};
