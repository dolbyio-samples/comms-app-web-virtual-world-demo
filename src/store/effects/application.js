import { createEffect } from "../utils/store";

export const sdkInitialized = createEffect("SDK_INITIALIZED");

export const setLocalParticipantID = createEffect("SET_LOCAL_PARTICIPANT_ID");
export const setCurrentConferenceID = createEffect("SET_CURRENT_CONFERENCE_ID");

export const setCurrentDeviceID = createEffect("SET_CURRENT_DEVICE_ID");

export const leaveConference = createEffect("LEAVE_CONFERENCE");

export const addMessageID = createEffect("ADD_MESSAGE_ID");
export const removeMessageID = createEffect("REMOVE_MESSAGE_ID");
