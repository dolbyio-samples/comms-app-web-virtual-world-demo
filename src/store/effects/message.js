import { createEffect } from "../utils/store";

export const addMessage = createEffect("ADD_MESSAGE");
export const removeMessage = createEffect("REMOVE_MESSAGE");
