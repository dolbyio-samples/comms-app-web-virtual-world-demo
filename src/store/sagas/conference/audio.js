import { call } from "redux-saga/effects";

import { getEntity } from "../../reducers/entities/selectors";
import {
  startLocalAudio as start,
  stopLocalAudio as stop,
} from "../../services/audio";
import { displayMessage } from "./message";

// FIXME: create a selectors files ?
export const getParticipant = (state, id) =>
  getEntity(state, "participants", id);

export function* startAudio() {
  const { error } = yield call(start);

  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not start audio",
    });
  }
}
export function* stopAudio() {
  const { error } = yield call(stop);

  if (error) {
    yield call(displayMessage, {
      type: "error",
      message: "could not stop audio",
    });
  }
}
