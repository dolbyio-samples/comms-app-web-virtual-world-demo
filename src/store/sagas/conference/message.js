import { createId } from "@paralleldrive/cuid2";
import { addMessage } from "../../effects/message";
import { application } from "../../effects";
import { put, call } from "redux-saga/effects";

export function* displayMessage({ type, message }) {
  const id = yield call(createId);
  yield put(
    addMessage({
      entityType: "messages",
      data: {
        [id]: {
          type,
          message,
        },
      },
    })
  );

  yield put(application.addMessageID({ id }));
}
