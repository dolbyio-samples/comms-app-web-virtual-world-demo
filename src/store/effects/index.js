import * as application from "./application";
import * as conference from "./conference";
import * as session from "./session";
import * as authentication from "./authentication";
import * as spatial from "./spatial";
import * as message from "./message";

export { authentication, application, conference, session, spatial, message };
