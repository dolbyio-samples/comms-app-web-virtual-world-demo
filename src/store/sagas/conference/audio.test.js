import { getParticipant, startAudio, stopAudio } from "./audio";
import { call } from "redux-saga/effects";
import {
  startLocalAudio as start,
  stopLocalAudio as stop,
} from "../../services/audio";
import { vi, describe, expect, it } from "vitest";
import { displayMessage } from "./message.js";

vi.mock("@voxeet/voxeet-web-sdk", () => ({
  default: {},
}));

describe("audio saga test suite", () => {
  it("should return a participant", () => {
    const state = {
      entities: {
        participants: {
          1: {
            name: "toto",
          },
        },
      },
    };

    const id = "1";

    const participant = getParticipant(state, id);

    expect(participant).toMatchSnapshot();
  });

  it("it should not call start audio service", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = startAudio(action);

    expect(generator.next().value).toEqual(call(start));
    expect(generator.next({ error: undefined }).done).toBeTruthy();
  });

  it("should call start audio service", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = startAudio(action);

    expect(generator.next().value).toEqual(call(start));
    expect(generator.next({ error: undefined }).done).toBeTruthy();
  });

  it("should call start audio service, fail and dispatch an action to the store ", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = startAudio(action);

    expect(generator.next().value).toEqual(call(start));
    expect(generator.next({ error: { undefined } }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not start audio",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });

  it("should call stop audio service", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = stopAudio(action);

    expect(generator.next().value).toEqual(call(stop));
    expect(generator.next({ error: undefined }).done).toBeTruthy();
  });

  it("should call stop audio service fail and dispatch an action to the store", () => {
    const action = { type: "START_AUDIO", payload: "123" };
    const generator = stopAudio(action);

    expect(generator.next().value).toEqual(call(stop));
    expect(generator.next({ error: {} }).value).toEqual(
      call(displayMessage, {
        type: "error",
        message: "could not stop audio",
      })
    );
    expect(generator.next().done).toBeTruthy();
  });
});
