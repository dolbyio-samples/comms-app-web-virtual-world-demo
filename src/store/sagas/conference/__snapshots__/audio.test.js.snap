// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`audio saga test suite > should return a participant 1`] = `
{
  "name": "toto",
}
`;
